pkg/dao/crd/dao.go

```diff
@@ -311,78 +315,163 @@ func (d *Dao) SetState(instanceID string, state apb.JobState) (string, error) {
+		return apb.JobState{}, err
+	}
+	for token, j := range bi.Status.Jobs {
+		// Assuming a single bind job happens per binding instance.
+		if j.Method == v1.JobMethodBind {
+			return apb.JobState{
+				Description: j.Description,
+				Method:      crd.ConvertJobMethodToAPB(j.Method),
+				Podname:     j.Podname,
+				Token:       token,
+				State:       crd.ConvertStateToAPB(j.State),
+				Error:       j.Error,
+			}, nil
+		}
+	}
+	return apb.JobState{}, &apierrors.StatusError{ErrStatus: metav1.Status{
+		Status: metav1.StatusFailure,
+		Code:   http.StatusNotFound,
+		Reason: metav1.StatusReasonNotFound,
+	}}
 }
 
 // FindJobStateByState - Retrieve all the jobs that match the specified state
 func (d *Dao) FindJobStateByState(state apb.State) ([]apb.RecoverStatus, error) {
-	log.Debugf("Dao::FindJobStateByState -> [%v]", state)
-	jobStates, err := d.client.JobStates(d.namespace).List(metav1.ListOptions{
-		LabelSelector: fmt.Sprintf("state=%v", crd.ConvertStateToCRD(state)),
-	})
+	sis, err := d.client.BundleInstances(d.namespace).List(metav1.ListOptions{})
 	if err != nil {
 		log.Errorf("unable to get job states for the state: %v - %v", state, err)
 		return nil, err
 	}
-
-	rs := []apb.RecoverStatus{}
-	errs := arrayErrors{}
-	for _, js := range jobStates.Items {
-		j, err := crd.ConvertJobStateToAPB(js.Spec, js.GetName())
-		if err != nil {
-			errs = append(errs, err)
-			continue
+	bis, err := d.client.BundleBindings(d.namespace).List(metav1.ListOptions{})
+	if err != nil {
+		log.Errorf("unable to get job states for the state: %v - %v", state, err)
+		return nil, err
+	}
+	rss := []apb.RecoverStatus{}
+
+	for _, si := range sis.Items {
+		for token, j := range si.Status.Jobs {
+			if state == crd.ConvertStateToAPB(j.State) {
+				rss = append(rss, apb.RecoverStatus{InstanceID: uuid.Parse(si.GetName()), State: apb.JobState{
+					Description: j.Description,
+					Method:      crd.ConvertJobMethodToAPB(j.Method),
+					Podname:     j.Podname,
+					Token:       token,
+					State:       crd.ConvertStateToAPB(j.State),
+					Error:       j.Error,
+				}})
+			}
 		}
-		rs = append(rs, apb.RecoverStatus{
-			InstanceID: uuid.Parse(js.GetLabels()[jobStateInstanceLabel]),
-			State:      *j,
-		})
 	}
-	if len(errs) > 0 {
-		return rs, errs
+	for _, bi := range bis.Items {
+		for token, j := range bi.Status.Jobs {
+			if state == crd.ConvertStateToAPB(j.State) {
+				rss = append(rss, apb.RecoverStatus{InstanceID: uuid.Parse(bi.GetName()), State: apb.JobState{
+					Description: j.Description,
+					Method:      crd.ConvertJobMethodToAPB(j.Method),
+					Podname:     j.Podname,
+					Token:       token,
+					State:       crd.ConvertStateToAPB(j.State),
+					Error:       j.Error,
+				}})
+			}
+		}
 	}
-	return rs, nil
+	return rss, nil
 }
 
 // GetSvcInstJobsByState - Lookup all jobs of a given state for a specific instance
 func (d *Dao) GetSvcInstJobsByState(ID string, state apb.State) ([]apb.JobState, error) {
```

Comment:
It's a little confusing seeing a "get service instance jobs by state" that also is looking for bindings? Does this imply that I can pass this a binding ID and it will give me back relevant jobs?

Reply:
Yes that is the case. I am making the functionality equivalent here but I agree with your point. I think this interface could use a second pair of eyes.